### DISCUSSION TITLE
Phields Bonsai infrastructure

### DISCUSSION THREAD

**Discussion by RoboDoig** [SainsburyWellcomeCentre/PhieldsExperiments#7] [General]
## Pattern discussions
- A common pattern in Aeon/Phields is modifying a HarpTimestamped value, e.g. a HarpTimestamped Aeon VideoDataFrame. For example, one may want to take a HarpTimestamped VideoDataFrame and convert the image to grayscale. Currently to do this the HarpTimestamped object must be 'split', the transformation applied and then the values zipped back together and timestamped. This ends up being a repeating pattern over alot of the workflows. Would be useful to have a special group workflow / operator that can do this without unzipping and rezipping every time. N.B. Doesn't this already exist with ConvertTimestamped?

- This can be even more generic than a HarpTimestamped operator. Could have a group type 'PropertyTransform' that is fed an observable sequence IObservable<T> and has a setting for x in T.x - the source of the group operates on x and returns IObservable<T> where T is modified. 

- Schemas for Phields initial experiments, need to design schemas that account for parameters involved in the initial experiments

## Triggering / State machine factory pattern
- Discussed with Andre a potential solution for combining schema selectable triggers and the state machine factory pattern. Produce a proposal structure that could be used for modular environment logic in Aeon/Phields

**Comment by aspaNeuro** [comment]
# Meeting Notes 06/05/2025
## Metadata
- Presented the current modular solution that consists in:
 - RigConfiguration Json schema where devices can be instantiated dynamically based on their family, creating a  list in yaml and using the device name to create a GroupBy with (device name in the key)
 - Devicename is defined by a enum this raises issues/advantages:
   - ADV - There is a dropdown, and you can never make a mistake with a device name 
   - ADV - Much easier to use in the workflow (for Logging/ControlPanel/Alerts/Tasks), less error prone invalidating at compiling time the Valid values 
    - ISS - More difficult to add new devices without regenerating the classes for bonsai thus stopping the workflow 
    - ISS - Generalisation can be more difficult and promoting such components to AEON core could lead to unwanted compromises with the device names.
    - Bruno Cruz is working on similar concepts and trying to create sparsed self contained Json see [issue](https://github.com/bonsai-rx/sgen/issues/40)
**side note** - the way we are setting up the exposure time may require refactoring to ensure property is setted up properly before acquisition starts and with no branching to avoid problems 
 - Discussion on modular encapsulation patterns with groupby stream names for event streams and command streams 
 - Each family of devices has a group workflow where the group by replay subjects that that family creates
## Logger
 - Presented Video Logger that creates different video files with different framerates and different folders and names 
## Task
 - Discussion on the trigger mapping design pattern for tasks 
## Control Panel 
 - Presented the concept of the main workflow working has a server and 
 - Discution on ZeroMQ / Zyre advertiser streams for enabling remote control pannels 
 - Decision to prototype this in a way that could be generic to try and find patterns 
## Data
 - Python data loader
 - Proposed using an extended HARP Python that deals with chunking automatically keeping the same standard of filenames that is currently being used by harp /python, although this idea seemed to tricky to implement so opting for a custom schema, the same way AEON is doing right now, was the final proposed solution.
 -  Same repo/branch that runs the acquisition should have the schema and Python environment to load data generated by the experimental workflow
## Modularization 
 - Discussion on what a module is in terms of granularity:
  1 - Each family of devices is a module that encapsulates: Events; Commands; Logging; Alerts; Control panel; Data schema module 
  2 - Each family has different modules the device module the control panel module alerts module ...
  3 - All logging is a module, All devices is a Module; All control panel is a module, All alerts is a module, All data loader is a module
